node/popn.go

```diff
@@ -799,12 +799,22 @@ func (nd *node) get(ctx context.Context, c cid.Cid, args *GetArgs) error {
 				return err
 			}
 		}
-		// Register new blocks in our supply by default
-		err = nd.exch.Index().SetRef(&exchange.DataRef{
-			PayloadCID:  c,
-			StoreID:     tx.StoreID(),
-			PayloadSize: int64(res.Size),
-		})
+
+		ref, err := nd.exch.Index().GetRef(c)
+		if err != nil {
+			ref = &exchange.DataRef{
+				PayloadCID:  c,
+				StoreID:     tx.StoreID(),
+				PayloadSize: int64(res.Size),
+				Keys:        [][]byte{},
+			}
+		}
+
+		if args.Key != "" {
```

Comment:
If `args.Key` == ""  we also assume that we just retrieved the whole thing so we're gonna need to get all the entries with MapKeys and set them